File: go/array/54_spiral_order.go
package array

// spiralOrder 螺旋矩阵
func spiralOrder(matrix [][]int) []int {
	// 模拟
	// 行，列 (m,n)
	rows, cols := len(matrix), len(matrix[0])

	visited := make([][]bool, rows)
	// 初始化visited 矩阵
	for i := 0; i < rows; i++ {
		visited[i] = make([]bool, cols)
	}

	var (
		// 矩阵的总长度
		total = rows * cols
		// 顺时针访问的顺序
		order    = make([]int, total)
		row, col = 0, 0
		// 索引访问的方向，用这个二维数组指示
		directions = [][]int{
			// 列+1， 向右
			{0, 1},
			// 行+1， 向下
			{1, 0},
			// 列-1， 下左
			{0, -1},
			// 行-1， 向上
			{-1, 0},
		}
		directionIdx = 0
	)

	for i := 0; i < total; i++ {
		order[i] = matrix[row][col]
		visited[row][col] = true

		nextRow, nextCol := row+directions[directionIdx][0], col+directions[directionIdx][1]
		// 判断下一次索引位置是否合法
		if nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols || visited[nextRow][nextCol] {
			// mod 4 是因为 directionIdx 一定在0-3 之间
			directionIdx = (directionIdx + 1) % 4
		}

		row += directions[directionIdx][0]
		col += directions[directionIdx][1]
	}
	return order
}

File: rust/src/array/54_spiral_order.rs

/// 螺旋矩阵
pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
    // 模拟
    let (rows, cols) = (matrix.len(), matrix[0].len());

    let mut visitd = vec![vec![false; cols]; rows];
    // 访问顺序的结果
    let total = rows * cols;
    let mut order = vec![0; total];
    // 右，下，左，上的顺序
    let directions: Vec<Vec<i32>> = vec![vec![0, 1], vec![1, 0], vec![0, -1], vec![-1, 0]];

    let mut direction_idx = 0;
    let (mut row, mut col) = (0, 0);

    for i in 0..total {
        order[i] = matrix[row][col];
        visitd[row][col] = true;

        let next_row = row as i32 + directions[direction_idx][0];
        let next_col = col as i32 + directions[direction_idx][1];
        // next_row, next_col 是否合法
        if next_row < 0
            || next_row >= rows as i32
            || next_col < 0
            || next_col >= cols as i32
            || visitd[next_row as usize][next_col as usize]
        {
            direction_idx = (direction_idx + 1) % 4;
        }
        // 经过上面的判断之后 row col as usize 是合法的
        row = (row as i32 + directions[direction_idx][0]) as usize;
        col = (col as i32 + directions[direction_idx][1]) as usize;
    }
    order
}
